### DISCUSSION TITLE
Trailing VarArgs

### DISCUSSION THREAD

**Discussion by ysndr** [pacak/bpaf#123] [Q&A] [answered]
Is it possible to catch trailing arguments such as

```
cmd sub --flag [-- ARGS...]
```

without adding some `Vec<String>` field to every subcommand?

**Comment by pacak** [comment]
Simple answer: it's easier to make them part of every command.

You will have to consume them within a command but this doesn't mean they have to be a part of a command structure: in combinatoric mode you can make it so every command returns a tuple of a command result itself and a vector of trailing arguments.

**Comment by ysndr** [comment]
Generally, a pair of `(parsed subcommand, anything else)` would be great

I hoped defining a struct 

```rust
struct Command {
	#[bpaf(external(sub_commands))]
	subcommand: SubCommands,

	#[bpaf(positional)]
	var_args: Vec<String>,
}
```

would allow that already since the help message says

```
Usage: COMMAND ... <ARG>...
```

**Reply by pacak** [reply]
Is it working? You can hide those positional arguments from the usage or help in general: <https://docs.rs/bpaf/latest/bpaf/trait.Parser.html#method.hide>

**Reply by pacak** [reply]
Hmm... I don't think that's going to work right now, sorry. Commands are a bit more special with them having their own help message and custom processing. The way I see the API - subcommand parser never returns.

**Reply by ysndr** [reply]
No its not working unfortunately, which again is a bit strange considering the help text

**Reply by ysndr** [reply]
any other way this could be made work without putting it in the subcommands?
I'm configuring a backend using these varargs and pass that to the processing of the subcommands, so it gets a bit strange when I had to process the var args there and reconfigure my backend stuff.

Possible to pull this one level higher? -> getting `(Command, Vec<String>)`

**Reply by pacak** [reply]
Let me make an example

**Reply by pacak** [reply]
https://github.com/pacak/bpaf/blob/complete_file/examples/shared_args.rs

**Reply by ysndr** [reply]
Thanks for the example
I'll have a look tomorrow if that solves my problem

**Comment by pacak** [comment, marked as answer]
You can also try to mark your command parser as `adjacent` https://pacak.github.io/bpaf/bpaf/trait.Parser.html#method.adjacent
